### PR TITLE
2 task

### PR DESCRIPTION
**Pull request by vetrslav**

### CODE REVIEW

**Comment by mkudosh** [comment]
**General notes:**

- Use strict comparison (=== instead of ==)

- Always add semicolons

- Use rest parameters instead of arguments object

- If your module exports just a single function, just export it (module.exports = yourFunction;)

- Don't implement useless duplicated tests. Instead of duplicating of test cases with different value, test edge cases

- Follow task specification carefully. If something is unclear get details from the customer (me)

**Implementation notes:**

- Helper.js

1. isOdd === !isEven. Define one function via another

2. Try to avoid using for loops and use built in array function to sum or multiply array

- PartialApplication.Test.js

1.  Confusing name randomSum. This variable holds partial application function, not random sum

2. Add test for case when partially applied function gets partially applied again (e.g. bind(bind(sum, 1), 2)(3) === 6)

3. Give more meaningful names to test cases. E.g. describe('Partial application to') it('sum with 1000 and argument 2 gives 1002')

- Currying.js

1. Implementation is not generic enough. Function should be applicable to any function with explicit parameters, not only to function with 3 parameters.

- Currying.Test.js

1.  Use simple test cases with numbers. Right now tests are too confusing.

- LinerUnfold.js

1. Don't use isWorking variable. It's name is not good and the not necessary: unfold should stop when callback returns falsy value.

2. Callback should return tuple with value and state for next iteration.

3. Unfold should accept not initial value, but initial state that will be passed to callback on first iteration. You shouldn't put initial state into resulting array.

- Map.js

1. Use linear fold instead of for loop.

- Filter.js

1. Use linear fold instead of for loop.

- Average.js

1. Use linear fold instead of for loop.

- Lazy.js

1. Lazy function should be evaluated only once.

- Lazy.Test.js

1. Ensure with your tests that function gets invoked only after lazy wrapper gets invoked (not when lazy wrapper is created).

2. Ensure with your tests that function gets invoked only once.

- Memoization.js

1. Function should have a single parameter.

2. Don't use arguments object as index in associative array.

- Memoization.Test.js

1. Ensure with your tests that function with a given argument gets invoked only once.

File: 2 TASK/js/Currying.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            return fn.apply(this,curryArgs)
+        }
+        else{
+            return  function() {
+                return curryRecursive.apply(this, slice.call(arguments, 0));
+            }
+        }
+    }
+    var curryArgs = [];
+    var ArgsCount = fn.length;
```
Comment (mkudosh) [inline 2 TASK/js/Currying.js]: Local variables should have lowercase names. Please fix.

File: 2 TASK/js/Lazy.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    return function () {
+        if (result) {
+            console.log("I remember this one!");
+            return result
+        }
+        result = lazyEval()
+        return result;
+    }
+}
+
```
Comment (mkudosh) [inline 2 TASK/js/Lazy.js]: This function should use memoization and save (argument, result) records in cache. Right now this function saves result of first call. If I call it second time with another argument it will return the same result as in first call.
